src/middlewared/middlewared/main.py

```diff
@@ -1370,8 +1370,19 @@ def terminate(self):
             # in base class to reduce number of awaits
             if hasattr(service, "terminate"):
                 self.logger.trace("Terminating %r", service)
+                terminate_timeout_func = getattr(service, 'terminate_timeout', None)
+                timeout = None
+                if terminate_timeout_func:
+                    if asyncio.iscoroutinefunction(terminate_timeout_func):
+                        timeout = await terminate_timeout_func()
+                    else:
+                        timeout = terminate_timeout_func()
```

Comment:
Someone will write synchronous I/O code there some day and it'll be executed in asyncio loop